### DISCUSSION TITLE
Will there be a new generator for plain async functions?

### DISCUSSION THREAD

**Discussion by capaj** [dotansimha/graphql-code-generator#8530] [Q&A] [answered]
Next.js 13 was released yesterday with:
https://nextjs.org/blog/next-13#data-fetching

If I want to use that, react-apollo-hooks generator cannot be used because I need plain async functions which return the data. Hooks won't work.

This going to need a new generator. Is there a plan to introduce it?

**Comment by dotansimha** [comment, marked as answer]
I guess ideally, you can use Codegen's `TypedDocumentNode` and have it integrated into your NextJS app (and the new async feature) with a small wrapper around the `fetch` function that hooks into `TypedDocumentNode`? 

(I'm also saying that in the context of preferring `TypedDocumentNode` in general, instead of framework-specific hooks)

**Reply by capaj** [reply]
I don't suppose you could share the code for that small wrapper around fetch so that it properly returns the type in the `TypedDocumentNode` or could you?

**Reply by n1ru4l** [reply]
@capaj You can find an example for Node.js using the HTTP module here: https://github.com/dotansimha/graphql-code-generator/blob/b4ac08ee7a8131cc983731232902a0f5db27f826/examples/typescript-esm/src/executeOperation.ts#L6-L34

Here is a `fetch` implementation.

```ts
import type { TypedDocumentNode } from '@graphql-typed-document-node/core';
import type { ExecutionResult } from 'graphql';
import { print } from 'graphql';

export function executeOperation<TResult, TVariables>(
  url: string,
  operation: TypedDocumentNode<TResult, TVariables>,
  ...[variables]: TVariables extends Record<string, never> ? [] : [TVariables]
): Promise<ExecutionResult<TResult>> {
  return fetch(url, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      query: print(operation),
      variables,
    }),
  }).then(res => res.json());
}
```

I recommend using the [`client` preset](https://www.the-guild.dev/graphql/codegen/docs/guides/react-vue) for generating the `TypedDocumentNodes`.
